### DISCUSSION TITLE
[V5] consume DateTime column type, toFormat is not function

### DISCUSSION THREAD

**Discussion by fmgono** [adonisjs/core#1876] [General]
I have a problem when i want to format date column,
here is the code :

in MtEmployee.ts
```
export default class MtEmployee extends BaseModel {
  /*...rest another columnt type*/

  @column.date({
    prepare: (value: string) => DateTime.fromFormat(value, 'dd/MM/yyyy'),
    consume: (value: DateTime) => value.toFormat('dd/MM/yyyy'),
  })
  public dateOfBirth: DateTime

  /*...rest another columnt type*/
}
```

but when i run this,response said toFormat is not function,
when i using function like `value.toLocaleString()`, its work, but that's not what i want...

i don't know why this happen, i think dateOfBirth is already in DateTime luxon object, so i can use some luxon function very well,

```
{
    "message": "value.toFormat is not a function",
    "stack": "TypeError: value.toFormat is not a function\n    at Object.consume (/home/fathan/fmgono/project/node-js/adonis-v5/how-my-employee-api/app/Models/Masters/MtEmployee.ts:61:41)\n    at /home/fathan/fmgono/project/node-js/adonis-v5/how-my-employee-api/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:1025:37\n    at Array.forEach (<anonymous>)\n    at Proxy.$consumeAdapterResult (/home/fathan/fmgono/project/node-js/adonis-v5/how-my-employee-api/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:1012:40)\n    at Function.$createFromAdapterResult (/home/fathan/fmgono/project/node-js/adonis-v5/how-my-employee-api/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:169:18)\n    at /home/fathan/fmgono/project/node-js/adonis-v5/how-my-employee-api/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:189:61\n    at Array.reduce (<anonymous>)\n    at Function.$createMultipleFromAdapterResult (/home/fathan/fmgono/project/node-js/adonis-v5/how-my-employee-api/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:187:31)\n    at ModelQueryBuilder.execQuery (/home/fathan/fmgono/project/node-js/adonis-v5/how-my-employee-api/node_modules/@adonisjs/lucid/build/src/Orm/QueryBuilder/index.js:112:43)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)"
}
```

can anyone help me with this ?

**Comment by thetutlage** [comment]
The `consume` method receives the value that is returned by the database and hence it is not a Luxon's DateTime instance. Also, I am not sure why you are dealing with string at all when the column type is set to `DateTime`.

If you can explain what exactly you are trying to achieve, then I maybe able to provide you some better suggestion

**Comment by sn0rcha** [comment]
I have the same problem here. I can't seem to get the date to format as dd/mm/yyyy on the consume.

**Reply by thetutlage** [reply]
Need to see code. I can't make assumptions and provide answers.

**Reply by sn0rcha** [reply]
Thanks for the reply. I'm really unsure how I should be handling this but here goes. Dates are always supplied from the SPA in dd/LL/yyyy (eg. 02/12/2020) - so I've got this in my code:

```
  @column.date({
    prepare: (value: string) => DateTime.fromFormat(value, 'dd/LL/yyyy'),
    consume: (value: string) => {
      DateTime.fromSQL(value).toFormat('dd/LL/yyyy')
    }
  })
```
I've tried ```DateTime.fromISO(value).toFormat('dd/LL/yyyy')``` as well but not matter what it's always having issues parsing with ```Invalid date/time```

The prepare works fine but not the consume on the way out.

**Reply by thetutlage** [reply]
What you are doing is wrong, since you should be dealing with the instance of luxon `DateTime` and not string. This is the reason you marked the column as `@column.date`. If you want to deal with strings directly, then you simply use the `@column()` decorator.

2nd what you need is string format during API communication and you should use the `serialize` method for that. This is how I will be doing in my apps.

```ts
class MyModel extends BaseModel {
  @column.date({
    // API response receives this value
    serialize: (value) => value && value.toFormat('dd/LL/yyyy')
  })
  public someColumn: DateTime
}
```

During creation. I will cast the request value to an instance of `Datetime`.

```ts
const model = new MyModel()
model.someColumn = DateTime.fromFormat(request.input('some_field'))

await model.save()
```

And finally, if you use the [validator](https://preview.adonisjs.com/guides/validator/schema-types#schemadate), it can handle the date format validation for you and also convert it to an instance of luxon DateTime as well.

**Reply by sn0rcha** [reply]
Thank you so much @thetutlage - that worked a treat!

**Comment by fmgono** [comment]
This is the right answer , Thank you very much

> What you are doing is wrong, since you should be dealing with the instance of luxon `DateTime` and not string. This is the reason you marked the column as `@column.date`. If you want to deal with strings directly, then you simply use the `@column()` decorator.
> 
> 2nd what you need is string format during API communication and you should use the `serialize` method for that. This is how I will be doing in my apps.
> 
> ```ts
> class MyModel extends BaseModel {
>   @column.date({
>     // API response receives this value
>     serialize: (value) => value && value.toFormat('dd/LL/yyyy')
>   })
>   public someColumn: DateTime
> }
> ```
> 
> During creation. I will cast the request value to an instance of `Datetime`.
> 
> ```ts
> const model = new MyModel()
> model.someColumn = DateTime.fromFormat(request.input('some_field'))
> 
> await model.save()
> ```
> 
> And finally, if you use the [validator](https://preview.adonisjs.com/guides/validator/schema-types#schemadate), it can handle the date format validation for you and also convert it to an instance of luxon DateTime as well.
